### PR TITLE
Design Database Interface

### LINKED ISSUE #13: Implement database

### PR DESCRIPTION
**Pull request by JavierMNieto**
Finishes #13. Since we will most likely want to batch data transfers, I am thinking of having a database transaction interface for each batch, so there will the main database connection which can have individual database transactions for committing data in batches.

### CODE REVIEW

**Review by alecchendev — COMMENTED** [review]
This is awesome. I'm learning a lot about how stuff is done in golang just by reading this code.

Just to get things clear, the `CSVDatabase` is separate from the `Database` interface right? I.e. `Database` is for taking csv files and uploading to Neo4j, while `CSVDatabase` is for taking "csvInserts" (blocks and txs) and writing to csvs (am I getting that right?).

Are you planning to test this next? Or are you going to implement other stuff within this PR. Just wondering.

**Comment by JavierMNieto** [comment]
`Database` is a general interface for interacting with any sort of Database similar to how `Client` is an interface for whatever node connection we implement. This way, we can easily select between different Database implementations we create (Neo4j, CSVs, SQL, etc.). So, `CSVDatabase` is just a general csv backend because while I was making the `Neo4jCSVDatabase` implementation I realized a lot of it was just handling writing to csvs so I separated it out into `CSVDatabase`.

 I wanted to get a basic implementation of `Neo4jCSVDatabase` done for our historical load into CSVs for Neo4j as well, but I am realizing I should prob do this in another PR.
 
Should I just push the `Database` interface stuff for now?

File: database/csv/csv.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		if err != nil {
+			return nil, err
+		}
+		db.csv_files.Store(name, csvFile)
+	}
+
+	// Start workers
+	for i := 0; i < maxWorkers; i++ {
+		g.Go(db.csvWorker)
+	}
```
Comment (alecchendev) [inline database/csv/csv.go]: Given the workers start in the instantiation. I'm assuming we'll interface with this by sending records to `db.msgs`? (As opposed to calling some function on db and passing in records) This seems to make sense for handling any new requests concurrently

(Mostly just comprehension check for myself)

**Comment by alecchendev** [comment]
> `Database` is a general interface for interacting with any sort of Database similar to how `Client` is an interface for whatever node connection we implement. This way, we can easily select between different Database implementations we create (Neo4j, CSVs, SQL, etc.). So, `CSVDatabase` is just a general csv backend because while I was making the `Neo4jCSVDatabase` implementation I realized a lot of it was just handling writing to csvs so I separated it out into `CSVDatabase`.
>
> I wanted to get a basic implementation of `Neo4jCSVDatabase` done for our historical load into CSVs for Neo4j as well, but I am realizing I should prob do this in another PR.

Gotcha. Yea I agree that makes sense to do in another PR.

> Should I just push the `Database` interface stuff for now?

Sure. I'll do a approve review.

**Review by alecchendev — APPROVED** [review]

File: database/csv/csv.go (hunk 1 above)
Comment (JavierMNieto) [reply database/csv/csv.go]: Yup exactly, I defined a `SendMsg` func [here](https://github.com/IlliniBlockchain/etl-bitcoin/pull/14/files#diff-66034a274276421c4c95008dc3eea5bc9fb472af7c76403afce8c9be501c1bb8R66) which the workers listen to [here](https://github.com/IlliniBlockchain/etl-bitcoin/pull/14/files#diff-66034a274276421c4c95008dc3eea5bc9fb472af7c76403afce8c9be501c1bb8R71)

**Pull request merged**